### PR TITLE
London | May-2025 | Halyna Kozlovska | Module Data Groups | Sprint 2

### PR DESCRIPTION
**Pull request by halyna-k**
## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with REGION | COHORT_NAME | FIRST_NAME LAST_NAME | PROJ_NAME
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [x] My changes meet the [requirements](./README.md) of this task

## Changelist

- Implemented multiple utility functions (contains, createLookup, tally, parseQueryString, invert) following the given requirements.
- Fixed bugs in initial implementations, including key handling, parsing edge cases, and dynamic property assignment.
- Improved error handling for invalid inputs.
- Handled edge cases: 
            - Empty objects or arrays
            - Duplicate keys
            - Keys without values
- Added comprehensive test coverage for all functions to ensure correctness, including edge case handling.
- Cleaned and optimized the code for readability and correctness.

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: Sprint-2/debug/address.js
Comment (fearcyf) [inline Sprint-2/debug/address.js:1]: works fine

File: Sprint-2/debug/author.js
Comment (fearcyf) [inline Sprint-2/debug/author.js:1]: works fine

File: Sprint-2/implement/lookup.test.js
Comment (fearcyf) [inline Sprint-2/implement/lookup.test.js:1]: I did the test in my IDE, it has passed. Well done

File: Sprint-2/implement/tally.test.js
Comment (fearcyf) [inline Sprint-2/implement/tally.test.js:1]: Test pass in my IDE, well done

Comment (halyna-k) [reply Sprint-2/implement/tally.test.js:1]: @fearcyf Thank you for your review.

Comment (halyna-k) [reply Sprint-2/implement/tally.test.js:1]: @fearcyf Do I need to change label to "Completed" or to leave "Need Review"?

File: Sprint-2/implement/contains.test.js
Comment (fearcyf) [inline Sprint-2/implement/contains.test.js:1]: 8 tests passed, one failed. This is the error message - >Error: Tests cannot be nested. Test "contains on empty object returns false" cannot run because it is nested within "given a contains function, should return true if the object contains the property, false otherwise".

Comment (fearcyf) [inline Sprint-2/implement/contains.test.js:1]: 9 out of 9 tests passed - well done. I have already mentioned in another review you can begin investigating how to make your tests more efficient and scalable.

**Comment by fearcyf** [comment]
If you need to discuss anything further feel free to message me.

**Comment by fearcyf** [comment]
@cjyuan  - I think we needs your approval before this submission can be finalised.

File: Sprint-2/implement/contains.test.js
Comment (halyna-k) [reply Sprint-2/implement/contains.test.js:1]: @fearcyf agree. Thank you.

**Comment by halyna-k** [comment]
> If you need to discuss anything further feel free to message me.

@fearcyf Thank you.

File: Sprint-2/implement/contains.test.js [hunk 1]
```diff
@@ -16,20 +16,58 @@ as the object doesn't contains a key of 'c'
...
+
+  // Given invalid parameters like an array
+  // When passed to contains
+  // Then it should return false or throw an error
+  [
+    { input: [[], "a"], expected: false },
+    { input: [null, "a"], expected: false },
+    { input: [undefined, "a"], expected: false },
+    { input: [123, "a"], expected: false },
+    { input: ["string", "a"], expected: false },
```
[the remark this answers is not in the document]
Comment (cjyuan) [reply Sprint-2/implement/contains.test.js:65]: I thought this test is to check for invalid cases. According to the comments at line 57-59:
>  // Given invalid parameters like an array
  // When passed to contains
  // Then it should return false or throw an error

Shouldn't  `contains([5, 5], 1)` returns `false`? `[5, 5]` is an array.

File: Sprint-2/implement/contains.js [hunk 2]
```diff
@@ -1,8 +1,13 @@
 function contains(obj, key) {
-  if (obj && typeof obj === "object") {
-    return Object.hasOwn(obj, key);
+  if (
+    obj === null ||
+    obj === undefined ||
```
Comment (cjyuan) [inline Sprint-2/implement/contains.js:4]: If obj is `undefined`, then the condition at line 5 would be false. So we could omit this check if we have the condition at line 5.

**Review by cjyuan — COMMENTED** [review]
Changes look good.
